### PR TITLE
Feature/blockchain generic test btc 214

### PR DESCRIPTION
**Pull request by blackyblack**

### CODE REVIEW

File: include/veriblock/blob.hpp [hunk 1]
```diff
@@ -40,8 +40,11 @@ struct Blob {
     data_.fill(0);
     assign(str);
   }
+  Blob(Blob<N>& other) {
```
Comment (Mr-Leshiy) [inline include/veriblock/blob.hpp]:
```suggestion
template<typename M>
  Blob(const Blob<M>& other) {
```

File: include/veriblock/blob.hpp [hunk 2]
```diff
@@ -40,8 +40,11 @@ struct Blob {
...
+    data_.fill(0);
```
Comment (Warchant) [inline include/veriblock/blob.hpp]: no need to do this before `assign`

File: include/veriblock/entities/vbkblock.hpp [hunk 3]
```diff
@@ -55,6 +59,30 @@ struct VbkBlock {
...
 
+  /*
+   * TODO
+   */
+  uint32_t getDifficulty() const;
+
+  /*
+   * TODO
+   */
+  uint32_t getBlockTime() const;
```
Comment (Warchant) [inline include/veriblock/entities/vbkblock.hpp]: here it is just a getter for `timestamp`

File: include/veriblock/entities/vbkblock.hpp [hunk 4]
```diff
@@ -55,6 +59,30 @@ struct VbkBlock {
    */
   void toVbkEncoding(WriteStream& stream) const;
 
+  /*
+   * TODO
+   */
+  uint32_t getDifficulty() const;
```
Comment (Warchant) [inline include/veriblock/entities/vbkblock.hpp]: it is just a getter for `difficulty`

File: test/blockchain/blockchain_test.cpp [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    EXPECT_CALL(*cursor, seekToFirst()).Times(1);
+    // database is empty, so first isValid will return false
+    EXPECT_CALL(*cursor, isValid()).WillOnce(Return(false));
+
+    // @when
+    auto result = blockchain->bootstrap(height, genesis, state);
+    EXPECT_TRUE(result);
+    EXPECT_TRUE(state.IsValid());
+  };
+};
```
Comment (Warchant) [inline test/blockchain/blockchain_test.cpp]: these two can be combined into single templated BlockchainInstance, and then parametrized with according types:
```
struct BtcTestCase {
  using params_t = BtcChainParams;
  using hash_t = decltype(BtcBlock::previousBlock);
  using block_t = BtcBlock;
  using height_t = int;
};

struct VbkTestCase {
  using params_t = VbkChainParams;
  using hash_t = decltype(VbkBlock::previousBlock);
  using block_t = VbkBlock;
  using height_t = int;
};
```

File: include/veriblock/blockchain/blocktree.hpp [hunk 6]
```diff
@@ -68,7 +68,9 @@ struct BlockTree {
       return state.addStackFunction("acceptBlockHeader()");
     }
 
-    auto it = block_index_.find(block.previousBlock);
+    // make a copy of hash to let it extend to hash_t size
+    hash_t previousBlockHash(block.previousBlock);
+    auto it = block_index_.find(previousBlockHash);
```
Comment (Warchant) [inline include/veriblock/blockchain/blocktree.hpp]: why is it needed?

**Comment by Warchant** [comment]
Overall looks good :)

File: include/veriblock/blockchain/blocktree.hpp (hunk 6 above)
Comment (blackyblack) [reply include/veriblock/blockchain/blocktree.hpp]: Auto conversion does not work here. Probably could be solved with proper Blob constructor but could not find it.

File: include/veriblock/entities/vbkblock.hpp (hunk 3 above)
Comment (blackyblack) [reply include/veriblock/entities/vbkblock.hpp]: Yes but we should keep it documented.

File: include/veriblock/blob.hpp (hunk 2 above)
Comment (blackyblack) [reply include/veriblock/blob.hpp]: Will it properly assign smaller size blob?

Comment (Warchant) [reply include/veriblock/blob.hpp]: Hm. I don't think so.

**Pull request merged**